### ISSUE TITLE
codepath of parallel processing

### ISSUE THREAD

**Issue by jasmainak** [jonescompneurolab/hnn-core#142]
While working on #110 [I realized](https://github.com/jonescompneurolab/hnn-core/pull/110/#discussion_r470928372) that the code path of the parallel processing is not quite linear and difficult to follow and read. It also matters if you want to give the user the ability to pass parameters to manipulate (e.g., use methods of) the `Network` class.

The current code path is as follows: 

- `simulate_dipole` instantiates the `_Backend` [which calls a method _BACKEND.simulate](https://github.com/jonescompneurolab/hnn-core/blob/master/hnn_core/dipole.py#L45)
- the `simulate` method in turn
  - for the `JoblibsBackend` calls [`_clone_and_simulate(net, ...)`](https://github.com/jonescompneurolab/hnn-core/blob/master/hnn_core/parallel_backends.py#L84) which calls `_simulate_single_trial(neuron_net)` that ultimately runs the simulation.
  - for the `MPIBackend` calls `mpi_child.py` which in turn calls `_simulate_single_trial(neuron_net)`

I think the way to improve it would be:

1. Make `mpi_child.py` use the same `_clone_and_simulate(net, ...)` rather than duplicating the code there so both backends have a common code path, thus reducing the chances of bugs due to differences in backends.
2. Add an abstract base class called `_BaseBackend` which includes the common methods so we enforce a similar structure. This will be inherited by both the backends.
3. Make `_simulate_single_trial` public so that power users can modify `neuron_net` as they like, then wrap it in their own parallel function (should be couple of lines with joblib). 

cc @blakecaldwell @rythorpe

**Comment by blakecaldwell** [comment]
In principle, I agree with 1 and 2 (more on that later). On 3, I don't see how a power user (skilled with NEURON-Python) would modify `neuron_net` with instantiated hoc objects and start a parallel simulation (which requires passing it to different processes). Do you have an idea?

Between the backends calling `__simulate_single_trial` , there is 1 intentional difference relating to the name `__clone_and_simulate` and why it is not used by the `MPIBackend`.

`neuron_net` is only instantiated once and reused by successive trials in `mpi_child.py`
https://github.com/jonescompneurolab/hnn-core/blob/master/hnn_core/mpi_child.py#L63

With `_clone_and_simulate`, each process instantiates it's own `neuron_net`. While necessary for joblibs, it would be inefficient for MPI. The current behavior should be maintained if the calls are combined.

Importantly, I think setting the seed should be made consistent between the two backends, arguing for a common function. The seed doesn't get set with the `MPIBackend` which really is a bug. This can be addressed in #113

**Comment by jasmainak** [comment]
yeah, this is what they would do:

```py
from hnn_core.neuron import NeuronNetwork, _simulate_single_trial

from joblib import Parallel, delayed

def _power_user_function(trial_idx):
    if trial_idx != 0:
         net.params['prng_*'] = trial_idx

    # any Neuron-Python code can also go here
    neuron_net = NeuronNetwork(net)
    neuron_net.do_whatever()  # this could be a method provided by our API

    return _simulate_single_trial(neuron_net). # hopefully this would be a public function

dpls = Parallel(n_jobs=1)(delayed(_power_user_function)(trial_idx) for i in range(10))
```

**Comment by jasmainak** [comment]
I am having some trouble wrapping my head around all the different functions though ... 

I was hoping that in the code above, instead of using `_simulate_single_trial`, I could use `simulate_dipole` but that takes as input an instance of `Network` not `NeuronNetwork`

**Comment by blakecaldwell** [comment]
> I am having some trouble wrapping my head around all the different functions though ...
> 
Maybe try modifying the current code and see how things fail? The issue is pickling.

> I was hoping that in the code above, instead of using `_simulate_single_trial`, I could use `simulate_dipole` but that takes as input an instance of `Network` not `NeuronNetwork`
Same reason for this--not having to pickle NEURON objects.

I see, you want the user to be able to define a custom function instead of `clone_and_simulate`. The base class could be documented with what needs to happen.
1. set any parameters
2. instantiate/build the `neuron_net` object
3. manipulate (neuron) network (e.g. change connectivity using a TBD API)
4. start simulation on (neuron) network

**Comment by jasmainak** [comment]
After working on #110 I feel that having a common `_clone_and_simulate` might be difficult. But wondering whether this can be part of the backend class or some such thing rather than loose code in `mpi_child.py`. I think it might be useful to separate the Neuron code from the MPI stuff as much as possible.

**Comment by blakecaldwell** [comment]
What do you think about this @jasmainak now that #296 is merged and #211 is closed? You were successful in creating a common `_clone_and_simulate`!

**Comment by jasmainak** [comment]
Let's close this issue!
